packages/widget-chat/src/components/activity-list/index.js

```diff
@@ -1,39 +1,78 @@
 import React, {PropTypes} from 'react';
 import classNames from 'classnames';
+import moment from 'moment';
+import {FormattedMessage} from 'react-intl';
 import ActivityItem from '../activity-item';
-
+import DaySeparator from '../day-separator';
+import ListSeparator from '../list-separator';
 import {formatDate} from '../../utils/date';
 
 export default function ActivityList(props) {
-  let lastActorId, lastVerb;
-  const {avatars, currentUserId, flags, onActivityDelete, onActivityFlag} = props;
-  const activities = props.activities
-    .map((activity) => {
-      const additional = lastActorId === activity.actor.id && lastVerb === activity.verb;
-      lastActorId = activity.actor.id;
-      lastVerb = activity.verb;
-      const isFlagged = flags && flags.some((flag) => flag.activityUrl === activity.url);
-      return (
-        <ActivityItem
-          activity={activity.object}
-          avatarUrl={avatars[activity.actor.id]}
-          id={activity.id}
-          isAdditional={additional}
-          isFlagged={isFlagged}
-          isSelf={currentUserId === activity.actor.id}
-          key={activity.id}
-          name={activity.actor.displayName}
-          onActivityDelete={onActivityDelete}
-          onActivityFlag={onActivityFlag}
-          timestamp={formatDate(activity.published)}
-          verb={activity.verb}
+  let lastActorId, lastDay, lastVerb;
+  const now = moment();
+  const {avatars, currentUserId, flags, lastAcknowledgedActivity, onActivityDelete, onActivityFlag} = props;
+  const items = [];
+  let shouldDisplayNewMessageMarker = false;
+  props.activities.forEach((activity) => {
```

Comment:
Doing this every time `activityList` renders/re-renders seems quite costly especially if we have a lot of items. Is there a way we can do this once, push it into into the store and just retrieve from there?